### PR TITLE
Add waterdrop monitoring

### PR DESCRIPTION
**Pull request by dannyd4315**
### What?

This PR adds monitoring for Waterdrop (producers) as part of the Karafka 2.0 update. There are a lot of changes in how things are instrumented in the Karafka update. So we are not able to track some of the older types of metrics. 

We have run through the metrics with @robertomiranda and agreed on the metric changes.

### Why?

In the main most of the metrics come from the statistics event that is emitted at a pre-determined time. (This will be set to every 5 seconds).

Other metrics come from specific event types using `ActiveSupport::Notifications`

### How?

The updated metrics that have been agreed between @robertomiranda, @alinagrishchuk and I are detailed below:

Waterdrop (Producer metrics)

## Async Producer Metrics (Global Web)

| Current metric | New Metric | Done |
| - | - | - |
| Queue Size | statistics/msg_cnt |  :white_check_mark: |
| Max queue size | statistics/msg_max |  :white_check_mark: |
| Queue fill ratio | Calculated queue size/max queue size |  :white_check_mark: |
| Buffer overflows | Error "producer" subscriber, tagged with "error name" |  :white_check_mark: |
| Dropped messages | Error "producer", tagged with "error name" |  :white_check_mark: |

## Fetcher Metrics

| Current metric | New Metric | Done |
| - | - | - |
| Fetcher queue size | N/A we already have this metric in consumer messages fetched | - |

## Producer Subscriber Metrics (Global Web)

| Current metric | New Metric | Done |
| - | - | - |
| Producer messages | statistics/msg_count |  :white_check_mark: |
| Producer message size | statistics/txmsg_bytes |  :white_check_mark: |
| Producer buffer size | N/A This is for async | - |
| Producer max buffer size | N/A This is for async |  - |
| Producer buffer fill ratio | N/A This is for async |  - |
| Producer buffer overflows | N/A This is for async |  - |
| Producer delivery errors | Error "producer", tagged with "error name"  | :white_check_mark: |
| Producer delivered messages | statistics/txmsgs |  :white_check_mark: |
| Producer delivery attempts | brokers/txretries | :white_check_mark: |
| Producer ack messages | N/A (We just care about errors) | - |
| Producer ack latency | statistics/rtt | :white_check_mark: |
| Producer ack errors | Parked for the moment | - |

Things to note: 

All errors are handled in the same way and increment a counter and add a label with the `error[:type]`.

### CODE REVIEW

**Comment by dannyd4315** [comment]
Note that all logging will be removed following review and testing before merge!

File: lib/promenade/waterdrop/error_subscriber.rb [hunk 1]
```diff
@@ -0,0 +1,27 @@
+require "promenade/waterdrop/subscriber"
+
+module Promenade
+  module Waterdrop
+    class ErrorSubscriber < Subscriber
+      attach_to "error.waterdrop"
+
+      Promenade.counter :waterdrop_errors do
```
Comment (robertomiranda) [inline lib/promenade/waterdrop/error_subscriber.rb:8]: Same naming suggestion as https://github.com/errm/promenade/pull/43/files#r1129265363
```suggestion
      Promenade.counter :waterdrop_errors_total do
```

File: lib/promenade/waterdrop/message_subscriber.rb [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+module Promenade
+  module Waterdrop
+    class MessageSubscriber < Subscriber
+      attach_to "message.waterdrop"
+
+      Promenade.counter :waterdrop_producer_messages do
+        doc "Number of messages written to Kafka producer"
+      end
+
+      Promenade.counter :waterdrop_producer_ack_messages do
```
Comment (robertomiranda) [inline lib/promenade/waterdrop/message_subscriber.rb:12]:
```suggestion
      Promenade.counter :waterdrop_producer_ack_messages_total do
```

File: lib/promenade/waterdrop/statistics_subscriber.rb [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+        def get_labels(statistics)
+          {
+            client: statistics[:client_id],
+          }
+        end
+
+        def convert_microseconds_to_seconds(time_in_microseconds)
+          time_in_microseconds / 1_000_000.to_f
+        end
```
Comment (robertomiranda) [inline lib/promenade/waterdrop/statistics_subscriber.rb:91]: nitpick this now has turned out into a pattern: 

https://github.com/errm/promenade/blob/2dd3ed334db8987e55e9e5c216415a5c74ee2f45/lib/promenade/karafka/statistics_subscriber.rb#L75-L77


what do you think about encapsulating this behaviour in a helper Object? perhaps in a separate PR

**Review by robertomiranda — APPROVED** [review]
LGTM 🚀 

Just one nitpick that could be addressed in separate PR 👍

**Pull request merged**